Sources/Flatten.swift

```diff
@@ -764,6 +786,126 @@ private struct LatestState<Value, Error: Swift.Error> {
 	var replacingInnerSignal: Bool = false
 }
 
+extension Signal where Value: SignalProducerProtocol, Error == Value.Error {
+	/// Returns a signal that forwards values from the "first input signal to send an event"
+	/// (winning signal) that is sent on `self`, ignoring values sent from other inner signals.
+	///
+	/// An error sent on `self` or the winning inner signal will be sent on the
+	/// returned signal.
+	///
+	/// The returned signal completes when `self` and the winning inner signal have both completed.
+	fileprivate func race() -> Signal<Value.Value, Error> {
+		return Signal<Value.Value, Error> { observer in
+			let composite = CompositeDisposable()
+			let relayDisposable = CompositeDisposable()
+
+			composite += relayDisposable
+			composite += self.observeRace(observer, relayDisposable)
+
+			return composite
+		}
+	}
+
+	fileprivate func observeRace(_ observer: ReactiveSwift.Observer<Value.Value, Error>, _ relayDisposable: CompositeDisposable) -> Disposable? {
+		let state = Atomic(RaceState<Value.Value, Error>())
+
+		return self.observe { event in
+			switch event {
+			case let .value(innerProducer):
+				// Ignore consecutive `innerProducer`s if any `innerSignal` already sent an event.
+				guard !relayDisposable.isDisposed else {
+					return
+				}
+
+				innerProducer.producer.startWithSignal { innerSignal, innerDisposable in
+
+					state.modify {
+						$0.innerSignalComplete = false
+					}
+
+					let disposableHandle = relayDisposable.add(innerDisposable)
+
+					innerSignal.observe { [unowned innerSignal] event in
+
+						let isWinningSignal: Bool = state.modify {
+							if $0.active == nil {
+								$0.active = innerSignal
+							}
+							return $0.active === innerSignal
+						}
+
+						// Ignore non-winning signals.
+						guard isWinningSignal else { return }
+
+						// Dispose all running innerSignals except winning one.
+						disposableHandle.remove()
```

Comment:
Is it possible to make these not run for every event? They just need to be run once after all.